datafusion/physical-plan/src/aggregates/group_values/column.rs

```diff
@@ -75,55 +154,350 @@ pub struct GroupValuesColumn {
-                | DataType::UInt8
-                | DataType::UInt16
-                | DataType::UInt32
-                | DataType::UInt64
-                | DataType::Float32
-                | DataType::Float64
-                | DataType::Utf8
-                | DataType::LargeUtf8
-                | DataType::Binary
-                | DataType::LargeBinary
-                | DataType::Date32
-                | DataType::Date64
-                | DataType::Utf8View
-                | DataType::BinaryView
-        )
+    /// Perform `vectorized_append`` for `rows` in `vectorized_append_row_indices`
+    fn vectorized_append(&mut self, cols: &[ArrayRef]) {
+        if self.vectorized_append_row_indices.is_empty() {
+            return;
+        }
+
+        let iter = self.group_values.iter_mut().zip(cols.iter());
+        for (group_column, col) in iter {
+            group_column.vectorized_append(col, &self.vectorized_append_row_indices);
+        }
+    }
+
+    /// Perform `vectorized_equal_to`
+    ///
+    /// 1. Perform `vectorized_equal_to` for `rows` in `vectorized_equal_to_group_indices`
+    ///    and `group_indices` in `vectorized_equal_to_group_indices`.
+    ///
+    /// 2. Check `equal_to_results`:
+    ///
+    ///    If found equal to `rows`, set the `group_indices` to `rows` in `groups`.
+    ///
+    ///    If found not equal to `row`s, just add them to `scalarized_indices`,
+    ///    and perform `scalarized_intern` for them after.
+    ///    Usually, such `rows` having same hash but different value with `exists rows`
+    ///    are very few.
+    fn vectorized_equal_to(&mut self, cols: &[ArrayRef], groups: &mut [usize]) {
+        assert_eq!(
+            self.vectorized_equal_to_group_indices.len(),
+            self.vectorized_equal_to_row_indices.len()
+        );
+
+        self.scalarized_indices.clear();
+
+        if self.vectorized_equal_to_group_indices.is_empty() {
+            return;
+        }
+
+        // 1. Perform `vectorized_equal_to` for `rows` in `vectorized_equal_to_group_indices`
+        //    and `group_indices` in `vectorized_equal_to_group_indices`
+        let mut equal_to_results = mem::take(&mut self.vectorized_equal_to_results);
+        equal_to_results.clear();
+        equal_to_results.resize(self.vectorized_equal_to_group_indices.len(), true);
+
+        for (col_idx, group_col) in self.group_values.iter().enumerate() {
+            group_col.vectorized_equal_to(
+                &self.vectorized_equal_to_group_indices,
+                &cols[col_idx],
+                &self.vectorized_equal_to_row_indices,
+                &mut equal_to_results,
+            );
+        }
+
+        // 2. Check `equal_to_results`, if found not equal to `row`s, just add them
+        //    to `scalarized_indices`, and perform `scalarized_intern` for them after.
+        let mut current_row_equal_to_result = false;
+        for (idx, &row) in self.vectorized_equal_to_row_indices.iter().enumerate() {
```

Comment:
Alternative way is to store the offset of the certain row instead of pushing equivalent row in `vectorized_equal_to_row_indices`. The benefit is we can reduce memory in low cardinality case which has many equality check.

Current:
vectorized_equal_to_row_indices: [row1, row1, row1, row2, row2, row2, row2]
vectorized_equal_to_group_indices: [gid1,gid2,gid3 ...]

Offset approach:
vectorized_equal_to_row_indices: [row1, row2]
vectorized_equal_to_offset: [0, 3, 7]
vectorized_equal_to_group_indices: [gid1,gid2,gid3 ...]

Reply:
Good point!
I plan to continue optimizing the `vectorized_equal_to` together with https://github.com/apache/datafusion/pull/12996#discussion_r1818601807 and https://github.com/apache/datafusion/pull/12996#discussion_r1823151542